### PR TITLE
Set sampling interval to UDATA_MAX to inform GC sampling is not required

### PR DESCRIPTION
**Pull request by JasonFengJ9**
**Set sampling interval to UDATA_MAX to inform GC sampling is not required**

GC need to be informed that the allocation sampling has been disabled, and turn off related GC settings;
Set sampling interval to `UDATA_MAX` in three cases:
1. `can_generate_sampled_object_alloc_events` capability is removed;
2. `JVMTI_EVENT_SAMPLED_OBJECT_ALLOC` event is disabled;
3. The `JVMTI` environment enabled `JVMTI_EVENT_SAMPLED_OBJECT_ALLOC` is disposed.

Related https://github.com/eclipse/openj9/issues/7740 https://github.com/eclipse/openj9/pull/9254 https://github.com/eclipse/omr/pull/5071

Reviewer: @DanHeidinga 
FYI: @LinHu2016 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

**Comment by JasonFengJ9** [comment]
`ci` build failure seems network related:
```
=== Output from failing command(s) repeated here ===
* For target jdk_modules_java.base__the.java.base_batch:
Connection attempt failed: Connection refused (Connection refused)
Connection attempt failed: Connection refused (Connection refused)
Connection attempt failed: Connection refused (Connection refused)
Giving up
IOException caught during compilation: Could not connect to server
```

**Comment by pshipton** [comment]
Restarted it.

File: runtime/jvmti/jvmtiEventManagement.c [hunk 1]
```diff
@@ -173,6 +173,14 @@ jvmtiSetEventNotificationMode(jvmtiEnv* env,
 					break;
  			}
 		}
+#if JAVA_SPEC_VERSION >= 11
+		else if (JVMTI_DISABLE == mode) {
+			if(JVMTI_EVENT_SAMPLED_OBJECT_ALLOC == event_type) {
```
Comment (DanHeidinga) [inline runtime/jvmti/jvmtiEventManagement.c:178]: nitpick
```suggestion
			if (JVMTI_EVENT_SAMPLED_OBJECT_ALLOC == event_type) {
```

File: runtime/jvmti/jvmtiCapability.c [hunk 2]
```diff
@@ -478,8 +478,10 @@ jvmtiRelinquishCapabilities(jvmtiEnv* env,
 
 #if JAVA_SPEC_VERSION >= 11
 		if (capabilities_ptr->can_generate_sampled_object_alloc_events) {
-			/* The default sampling interval is not changed. */
+			J9VMThread *currentThread = vm->internalVMFunctions->currentVMThread(vm);
```
Comment (DanHeidinga) [inline runtime/jvmti/jvmtiCapability.c:481]: There's a `getCurrentThread` jvmti helper method to handle unattached threads correctly.  It should probably be used here as well.

File: runtime/jvmti/jvmtiHelpers.c [hunk 3]
```diff
@@ -174,13 +175,14 @@ disposeEnvironment(J9JVMTIEnv * j9env, UDATA freeData)
 #if JAVA_SPEC_VERSION >= 11
 		if (j9env->capabilities.can_generate_sampled_object_alloc_events) {
 			J9JVMTI_DATA_FROM_VM(vm)->flags &= ~J9JVMTI_FLAG_SAMPLED_OBJECT_ALLOC_ENABLED;
+			/* Set sampling interval to UDATA_MAX to inform GC that sampling is not required */
+			vm->memoryManagerFunctions->j9gc_set_allocation_sampling_interval(currentThread, UDATA_MAX);
```
Comment (DanHeidinga) [inline runtime/jvmti/jvmtiHelpers.c:179]: Do we need some way to push this out to all the threads? (ie: an async event?) Is this event per thread or global?

File: runtime/jvmti/jvmtiEventManagement.c (hunk 1 above)
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiEventManagement.c]: fixed.

File: runtime/jvmti/jvmtiCapability.c (hunk 2 above)
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiCapability.c]: Agreed, use the helper method `getCurrentVMThread` here instead. 
For another usage of `currentThread` at `jvmtiHelpers.c:disposeEnvironment()` without a return error code, check if the `currentThread` is `NULL` directly, and skip the use of `currentThread` when it is unattached and has a `NULL` value.

File: runtime/jvmti/jvmtiHelpers.c (hunk 3 above)
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiHelpers.c]: If I understand the question correctly, there is no need to do so for all threads.
`j9gc_set_allocation_sampling_interval(vmThread, samplingInterval)` modifies `MM_GCExtensions.oolObjectSamplingBytesGranularity`
https://github.com/eclipse/omr/blob/eb71465140b3b7fd74f8ad12a65da4d384aa945e/gc/base/GCExtensionsBase.hpp#L295
which is a global setting for GC to determine if the newly added `MM_GCExtensions. _traceAllocationBytes` (https://github.com/eclipse/omr/pull/5071/files) has passed the `samplingInterval` and a sampling event should be fired.

`can_generate_sampled_object_alloc_events` is global, the sampled objects come from the allocations in different threads.

File: runtime/jvmti/jvmtiCapability.c [hunk 4]
```diff
@@ -478,8 +478,13 @@ jvmtiRelinquishCapabilities(jvmtiEnv* env,
 
 #if JAVA_SPEC_VERSION >= 11
 		if (capabilities_ptr->can_generate_sampled_object_alloc_events) {
-			/* The default sampling interval is not changed. */
+			J9VMThread *currentThread = NULL;
 			jvmtiData->flags &= ~J9JVMTI_FLAG_SAMPLED_OBJECT_ALLOC_ENABLED;
+			rc = getCurrentVMThread(vm, &currentThread);
+			if (JVMTI_ERROR_NONE == rc) {
+				/* Set sampling interval to UDATA_MAX to inform GC that sampling is not required */
+				vm->memoryManagerFunctions->j9gc_set_allocation_sampling_interval(currentThread, UDATA_MAX);
```
Comment (gacholio) [inline runtime/jvmti/jvmtiCapability.c:486]: If the sampling interval is global, the GC call should be changed to take a VM, not a thread.

Comment (gacholio) [reply runtime/jvmti/jvmtiCapability.c:486]: @dmitripivkine Any objection?

Comment (dmitripivkine) [reply runtime/jvmti/jvmtiCapability.c:486]: Seems right for me. Is there any thread-specific usage of this API I missed?

Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiCapability.c:486]: > Is there any thread-specific usage of this API I missed?

No, the `currentThread` is used to retrieve `J9JavaVM` instead.

Comment (LinHu2016) [reply runtime/jvmti/jvmtiCapability.c:486]: > Seems right for me. Is there any thread-specific usage of this API I missed?

we need MM_EnvironmentBase to retrieve _objectAllocationInterface for disableCachedAllocations() in https://github.com/eclipse/openj9/pull/9254

Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiCapability.c:486]: > we need MM_EnvironmentBase to retrieve _objectAllocationInterface for disableCachedAllocations() in #9254

Post a question at https://github.com/eclipse/openj9/pull/9254/files#r409765714

Comment (LinHu2016) [reply runtime/jvmti/jvmtiCapability.c:486]: Maybe we can retrieve MM_EnvironmentBase via getCurrentVMThread(), or somewhere else

Comment (dmitripivkine) [reply runtime/jvmti/jvmtiCapability.c:486]: There is not a question how to get thread info, if we really need it we can keep API as it stays now. However there is a question about correctness global vs per thread. Is it correct to set is as now for thread or we should manage to set this for ALL threads?

Comment (gacholio) [reply runtime/jvmti/jvmtiCapability.c:486]: My complaint here was all the work to retrieve the thread, only to have it used to fetch the VM.

Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiCapability.c:486]: `j9gc_set_allocation_sampling_interval()` is intended to affect all threads. If https://github.com/eclipse/openj9/pull/9254 can find a way for `disableCachedAllocations()` via `J9JavaVM *` instead of `J9VMThread *`, `j9gc_set_allocation_sampling_interval()` can be modified to take `J9JavaVM *` which gives clear indication that this API is for VM all threads.

Comment (dmitripivkine) [reply runtime/jvmti/jvmtiCapability.c:486]: I believe this is true. However current implementation in https://github.com/eclipse/openj9/pull/9254 sets flag for one thread only. It does not look correct for me. We need to double check. And it confirms GAC position - it is useful to set correct scope

Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiCapability.c:486]: Okay, am going to change `j9gc_set_allocation_sampling_interval()` to take `J9JavaVM *`.

@LinHu2016 please adjust your PR & personal build.

Comment (LinHu2016) [reply runtime/jvmti/jvmtiCapability.c:486]: sure, I am going to handle whole threads for disableCacheAllocations and monitor extensions->disableInlineAllocationForSamplingBytesGranularity for initializing _cachedAllocationsEnabled for new created thread, I am going to leave j9gc_set_allocation_sampling_interval() no change until this PR is merged.

Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiCapability.c]: Updated the PR to accept `J9JavaVM *` in `j9gc_set_allocation_sampling_interval()`. Keep the `NULL` check for `currentThread` within `disposeEnvironment()`. `deleteAgentBreakpoint(J9VMThread *, J9JVMTIEnv *, J9JVMTIAgentBreakpoint *)` and its associates might not need `J9VMThread *currentThread`, but will leave that to another PR.

@gacholio @DanHeidinga please review.

File: runtime/jvmti/jvmtiHelpers.c [hunk 5]
```diff
@@ -174,20 +174,25 @@ disposeEnvironment(J9JVMTIEnv * j9env, UDATA freeData)
...
+		if (NULL != j9env->breakpoints) {
 			J9VMThread * currentThread = vm->internalVMFunctions->currentVMThread(vm);
-			pool_state poolState;
-			J9JVMTIAgentBreakpoint * agentBreakpoint = NULL;
-
-			agentBreakpoint = pool_startDo(j9env->breakpoints, &poolState);
-			while (agentBreakpoint != NULL) {
-				deleteAgentBreakpoint(currentThread, j9env, agentBreakpoint);
-				agentBreakpoint = pool_nextDo(&poolState);
+			/* Ignore the case that currentThread is NULL, i.e., unattached thread */
```
Comment (gacholio) [inline runtime/jvmti/jvmtiHelpers.c:186]: Why is this necessary?

Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiHelpers.c:186]: `currentThread` is from `vm->internalVMFunctions->currentVMThread(vm)` which might return `NULL` as per https://github.com/eclipse/openj9/blob/5800af9452097b029c1a38fbc5cbd20011060a8b/runtime/jvmti/jvmtiHelpers.c#L502-L510
Current function `disposeEnvironment()` doesn't return an error code, hence just call `deleteAgentBreakpoint(currentThread, j9env, agentBreakpoint)` when the `currentThread` is not`NULL`, and ignore the `NULL` case.

Comment (gacholio) [reply runtime/jvmti/jvmtiHelpers.c:186]: Is this to fix something? I don't believe this situation could actually occur, so I'd rather see the crash than a silent no-op.

Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiHelpers.c:186]: No, this change is completely from eyeballing the code. Will restore the code then.

File: runtime/jvmti/jvmtiMemory.c [hunk 6]
```diff
@@ -87,19 +87,15 @@ jvmtiSetHeapSamplingInterval(jvmtiEnv *env,
...
 	ENSURE_CAPABILITY(env, can_generate_sampled_object_alloc_events);
 	ENSURE_NON_NEGATIVE(samplingInterval);
 
-	rc = getCurrentVMThread(vm, &currentThread);
-	if ((JVMTI_ERROR_NONE == rc) && (NULL != currentThread)) {
-		/* No negative samplingInterval, and there is no data lost when jint is casted to UDATA. */
-		vm->memoryManagerFunctions->j9gc_set_allocation_sampling_interval(currentThread, samplingInterval);
-	}
+	/* No negative samplingInterval, and there is no data lost when jint is casted to UDATA. */
+	currentThread->javaVM->memoryManagerFunctions->j9gc_set_allocation_sampling_interval(((J9JVMTIEnv *)env)->vm, samplingInterval);
```
Comment (gacholio) [inline runtime/jvmti/jvmtiMemory.c:98]: VM is fetched twice here, in two different ways. Please use a local.

Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiMemory.c]: Fixed, please have another look.

**Comment by gacholio** [comment]
jenkins test sanity zlinux jdk11
